### ISSUE TITLE
Proposal: Allow a fixer to apply multiple fixes simultaneously

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#7348]
## Problem

As of ESLint 3.7, the autofixer for a rule can only perform one "fix" at a time. In this context, a "fix" is an atomic change (removal/insertion/replacement) to a single contiguous range of code.

From what I understand, this restriction was introduced when the autofixers for core rules almost exclusively fixed whitespace, and didn't touch any tokens. The intention was to ensure that each fix was as small as possible, to increase the likelihood of any given fix getting applied.

There are cases where a rule wants to apply multiple fixes at separate locations. Nowadays, many autofixers modify tokens and the AST, not just whitespace. As a result, these multi-fix operations cannot be split into smaller atomic operations, because the syntax or runtime behavior of the code will be incorrect if only one fix gets applied without the other.

The autofixer for [`no-extra-parens`](http://eslint.org/docs/rules/no-extra-parens) is a good example of this:

``` js
/* eslint no-extra-parens: error */

var foo = (function () {
  // function body
});
```

The autofixer for `no-extra-parens` would like to remove the parentheses around the function, to obtain the following code:

``` js
var foo = function () {
  // function body
};
```

Clearly, the left and right parentheses have to be removed in the same pass of the code; if only one paren is removed without the other, the resulting syntax will be invalid. Reporting two separate errors as a workaround does not solve this issue; since the fix from any given reported error is not guaranteed to be applied, reporting an for each paren can still result in only one fix being applied under certain circumstances.

Due to these restrictions, `no-extra-parens` (and many other core rules, such as `prefer-arrow-callback`, `quote-props`, `wrap-regex`, etc.) uses a a different workaround: it replaces the entire text range between the two parens. Applied to this example, it replaces the entire range containing `(function () { ... })` with `function () { ... }`.

This workaround works because it only applies to a single text range, so it is a single contiguous fix. However, it is not ideal. Since the resulting fix is so large, it can prevent other fixes from being applied during the same pass; this results in a larger number of passes, which hurts the performance of autofixing. The fix is also overcomplicated intuitively; for a large function, it could easily end up replacing tens of thousands of characters, even though it is only trying to remove two characters.
## Proposal: Allow multiple fixes

This problem could be solved if a rule could apply multiple fixes per reported problem. For example, the `no-extra-parens` autofixer would apply two fixes; a single-character fix to remove the left paren, and a single-character fix to remove the right paren.

The fixes for a given reported problem would be applied atomically. In other words, there would still be no guarantee that any problem's fixes would get applied in general, but the core engine would guarantee that all the fixes for a problem would be applied simultaneously, so the code would never end up in a state where only some of a problem's fixes were applied.
### API Changes

Currently, fixes are applied when a fixer function returns an object with `range` and `text` keys. (Usually, these fix objects are generated by helper functions such as `RuleFixer#replaceText`.)

To apply multiple fixes, a fixer function could instead return an _array_ of these fix objects. Returning a single fix object would still be supported for backwards-compatibility.

For example, the fixer function for `no-extra-parens` would look something like this:

``` js
context.report({
  node,
  message,
  fix(fixer) {
    return [
      fixer.remove(leftParenToken),
      fixer.remove(rightParenToken)
    ];
  }
});
```

**Comment by platinumazure** [comment]
I'm just going to leave one example showing why this would be somewhat difficult:

Before:

```js
const sum = (5+8);
```

Rules being applied:
* space-infix-ops (should add spaces to each side of + token)
* no-extra-parens (remove the parentheses)

Desired after:

```js
const sum = 5 + 8;
```

Fixes to be applied:

```js
[
    // no-extra-parens: remove parentheses
    [{ range: [12, 13], text: "" }, { range: [18, 19], text: "" }],

    // space-infix-ops: replace operator with padded operator
    [{ range: [14, 15], text: " + " }]
]
```

The current implementation is efficient because fixes are applied in reverse order of starting range. This means we iterate through fixes in linear time and we do not have to alter the ranges of any fixes (because a fix later in the text does not affect the location of a fix earlier in the source).

A naive implementation that focuses on atomicity would have to iterate through the fix queue and potentially update locations for remaining fixes, so this would become `O(n^2)` without very clever thinking. In the example above, this applies no matter which rule's fixer you apply first:

* space-infix-ops first: We need to ensure the range for the second fix of no-extra-parens has its location adjusted.
* no-extra-parens first: We need to ensure the range for the space-infix-ops fix has its location adjusted.

I'm sure nzakas will have many more examples. I just wanted to make sure the size of the problem (which I may still be under-representing) is clear.

**Comment by not-an-aardvark** [comment]
Interesting example. I'll think about whether it's possible to maintain linear time and atomicity, but I'm also wondering: How bad for overall performance would a quadratic-time implementation actually be?

For example, if a file has 100 fixes (probably much larger than the average), this requires updating ranges `O(10000)` times. I'm inclined to think that updating a range is computationally inexpensive, as it just requires a couple arithmetic operations, so even 10000 range updates would not cause a performance difference noticeable to the user. In comparison, doing another autofixing pass is very expensive, as it requires reparsing the file and running every rule again on the new AST.

In other words, even if this change would require the autofixing implementation to be O(n<sup>2</sup>) in the number of fixes, it might still improve performance overall by reducing the number of passes required (since fewer fixes would conflict with each other)

**Comment by not-an-aardvark** [comment]
Here's a linear-time implementation:

1. Create a linked list of character blobs in the file. A character blob is a range of characters that is not split by any fixes. For n fixes, there will be at most 2n+1 blobs.
1. Match each fix to a starting blob and ending blob, based on the ranges of the fix.
1. Apply each fix by replacing the appropriate blobs for a given fix with a new blob for the replacement text.
1. Reassemble the linked list of blobs into a string

**Comment by nzakas** [comment]
Related: https://github.com/eslint/eslint/issues/5329

If I'm understanding you correctly, it sounds like in the case of an autofix that needs to apply two changes, you'll apply the first one, then apply any intervening fixes while keeping an offset of the fixes already made so that the second one can be made in the correct location?

**Comment by not-an-aardvark** [comment]
Not quite: I was saying I would store the text of the file as a linked list and convert all the ranges appropriately at the start. That way, the fixer wouldn't have to deal with updating ranges and offsets, and it could still be O(n).

Although now that you mention it, keeping an offset might also allow for O(n) time while keeping the solution simpler overall.

Not to deter from the implementation discussion, but what do you think of the proposed API change itself? If the proposal ends up getting accepted, then I can draft an implementation which might make discussion about it easier.

**Comment by platinumazure** [comment]
I'm okay with the API. I just want to make sure we agree on what the requirements of a new implementation should be (which may include things like linear time performance, atomicity of sets of fixes, etc.)

**Comment by not-an-aardvark** [comment]
Atomicity should definitely be required for all the fixes from a given reported problem. (A fix for removing parens, such as the one in the original comment, is only safe if the fixes are atomic).

I think linear-time performance should also be a requirement for the implementation, provided that linear-time performance is possible to achieve. I'm pretty sure it is possible, but if we discover that I'm wrong and it's impossible, then I think it would still be worth implementing to determine the actual performance impact of a quadratic-time implementation.

**Comment by not-an-aardvark** [comment]
Slight correction: The existing implementation appears to be `O(n log n)` time, not linear time (since the list of fixes is sorted beforehand). So the new implementation should be at most `O(n log n)` time.

**Comment by not-an-aardvark** [comment]
Sample `O(n log n)` implementation: https://github.com/eslint/eslint/commit/48357ea3d2631512584a07166dce5c88c8361c15

It ended up being simpler than I was describing it; the fixer now does the check for conflicts before any fixing takes place. This allows allows it to "undo" a fix by simply removing it from the list of fixes to apply if it later discovers a conflict on the same reported problem. After checking for conflicts and creating a list of fixes that can be applied without conflicts, the fixer goes ahead and applies the fixes.

**Comment by nzakas** [comment]
The API itself is fine by me, I'm much more worried about the implementation, and most specifically, that we won't end up doing fixes incorrectly. The primary reason I required just one change per message was to ensure the correctness of the output. For instance, if one rule removes a comma while another tries to insert a space before a comma, if the comma removal happens before the space insertion, we've now altered the text modification and might have inserted a space in the middle of an identifier.

How do you ensure that multiple fixes for a message are always applied and that fixes that might apply in between don't interfere?

(note: I'm not nearly as worried about performance as I am about correctness)

**Comment by not-an-aardvark** [comment]
Interesting question. I was going to make an argument involving context-free grammars, but then I realized that it's moot because JS isn't completely context-free anyway. As a result, there are a few cases where non-overlapping fixes can conflict with each other, even with the current behavior.

---

Example:

Rule A is a custom rule that replaces all non-default function parameters with a default parameter of `undefined`. For example, it performs this fix:

```js
function foo(param) {}

// Rule A fixes this to:

function foo(param = undefined) {}
```

Rule B is like the opposite of ESLint's `strict` rule; it adds redundant `'use strict';` directives to functions

```js
'use strict';
function foo(param) {}

// Rule B fixes this to:

'use strict';
function foo(param) {'use strict';}
```

(Granted, I'm not sure why anyone would actually want to use either of these rules, but bear with me.)

Suppose rules A and B are applied to the following code, in autofix mode:

```js
'use strict';

function foo(param} {}
```

Rule A will replace `param` with `param=undefined`, and Rule B will insert `'use strict';` in the function body. Individually, each of the fixes can be applied without changing the semantics of the code. These fixes do not overlap, so both of them will be applied by the current version of the fixer:

```js
'use strict';

function foo(param=undefined) {'use strict'}
```

But this is a SyntaxError, since default parameters aren't allowed in functions with `'use strict';` directives.

---

Admittedly, this doesn't really answer your question, but I thought it was worth pointing out that the existing behavior isn't correct 100% of the time either.

**Comment by alberto** [comment]
I was intending to try implement something like this for a while, I think it's doable with the current design.

We apply non-overlapping fixes from the bottom up. With this approach a fix would consist of a set of multiple ranges and the whole set would be applied on a single pass. Other fixes would only be applied if their ranges didn't overlap with the sets of fixes already applied (or queued for applying in that pass). Everything else would work as usual. Makes sense?

**Comment by gibson042** [comment]
I think you want [Operational Transformation](https://en.wikipedia.org/wiki/Operational_transformation), and in particular https://www.npmjs.com/package/ot .

**Comment by not-an-aardvark** [comment]
@gibson042 Sorry, I don't understand; could you clarify how OT is useful in making sure no syntax errors are produced?

**Comment by gibson042** [comment]
It can't guarantee that (as you pointed out above, even current application can generate them). But it will allow atomic application of multiple ranges and conflict detection/resolution without reinventing the wheel, addressing @nzakas concerns at https://github.com/eslint/eslint/issues/7348#issuecomment-254300563 and https://github.com/eslint/eslint/issues/7348#issuecomment-255441028 .

**Comment by alberto** [comment]
@nzakas I was replying to everyone in the sense that I think my implementation proposal is not the same as @not-an-aardvark if I understood him correctly.

My proposal is not very different from what we have now, only adapted to multiple changes. Let me explain the process in more detail:
- a fix can consist of one or more changes (right now, one fix === one change). The same api @not-an-aardvark proposed in the first comment.
- We sort fixes in descending change order (by their highest change range). We take the first fix in the list (the one with the highest change range) and apply that change, since there are no possible conflicts yet. We don't apply the rest of the changes in this fix immediately, but store them to apply them in the future, but in the same fix iteration (before we lint the file again, hence the "atomicity"). Let's say these changes have been accepted, for lack of a better term.
- We take the next fix from the list. If its changes don't overlap with any of the changes already accepted, the last change of this fix is applied and the rest of the changes are added to the accepted list. If any of the changes in this fix overlaps with the already accepted changes, the fix is not applied.
- Repeat for every remaining fix.

Since changes are applied in reverse order, as they are currently done, there is no need for an offset or updating ranges.

In the example @platinumazure [mentioned](https://github.com/eslint/eslint/issues/7348#issuecomment-253328207), the order in which the changes would be applied would be:
[18,19], [14,15], [12, 13]

**Comment by not-an-aardvark** [comment]
Marking as accepted because https://github.com/eslint/eslint/issues/8101 was accepted in the [TSC meeting](https://github.com/eslint/tsc-meetings/blob/01a46b79c419dd853dfdeda676ab9969ffbcd2c2/notes/2017/2017-05-11.md) a few days ago.
